File: controller/evdev_surface.py
# SPDX-License-Identifier: AGPL-3.0-only WITH OzmaPluginException
"""
Generic evdev control surface driver for ozma.

Supports any Linux input device that exposes buttons (EV_KEY) and/or
axes (EV_ABS, EV_REL) — this covers:

  - Contour ShuttlePRO / ShuttleXpress (jog + shuttle + buttons)
  - USB foot pedals (single or multi-pedal)
  - Macro keypads (Razer Tartarus, Logitech G13, etc.)
  - USB button boxes / switch panels
  - Any other /dev/input device not caught by the gamepad driver

Config-driven: each button or axis is mapped to an ozma action in
controls.yaml.  The driver matches devices by name substring.

Example controls.yaml entry::

    surfaces:
      shuttle:
        type: evdev
        device: "ShuttlePRO"          # Substring match on device name
        grab: true                     # Exclusively grab the device
        buttons:
          260: { action: scenario.next, value: -1 }   # Button 1 → prev
          261: { action: scenario.next, value: 1 }     # Button 2 → next
          262: { action: audio.mute, target: "@active" }
        axes:
          7: { action: audio.volume, target: "@active", min: 0, max: 255 }
"""

from __future__ import annotations

import asyncio
import logging
from typing import Any

import evdev
from evdev import InputDevice, ecodes

from controls import ControlSurface, Control, ControlBinding

log = logging.getLogger("ozma.evdev_surface")


def find_evdev_device(name_pattern: str) -> InputDevice | None:
    """Find an evdev device whose name contains the given pattern (case-insensitive)."""
    pattern_lower = name_pattern.lower()
    for path in evdev.list_devices():
        try:
            dev = InputDevice(path)
            if pattern_lower in dev.name.lower():
                return dev
        except Exception:
            pass
    return None


class EvdevSurface(ControlSurface):
    """
    A generic evdev input device registered as an ozma control surface.

    Config structure::

        {
            "device": "ShuttlePRO",   # Name substring match
            "grab": true,             # Exclusive grab
            "buttons": {
                "260": {"action": "scenario.next", "value": -1},
                ...
            },
            "axes": {
                "7": {"action": "audio.volume", "target": "@active"},
                ...
            },
            "rel_axes": {
                "7": {"action": "scenario.next"},  # REL_DIAL on ShuttlePRO
                ...
            }
        }
    """

    def __init__(self, surface_id: str, config: dict) -> None:
        super().__init__(surface_id)
        self._config = config
        self._device_pattern = config.get("device", "")
        self._grab = config.get("grab", False)
        self._device: InputDevice | None = None
        self._task: asyncio.Task | None = None
        self._on_changed: Any = None

        # Parse button mappings: evdev_code → ControlBinding
        self._button_map: dict[int, tuple[str, ControlBinding]] = {}
        for code_str, binding_cfg in config.get("buttons", {}).items():
            code = int(code_str)
            name = f"btn_{code}"
            binding = ControlBinding(
                action=binding_cfg.get("action", ""),
                target=binding_cfg.get("target", ""),
                value=binding_cfg.get("value"),
            )
            self._button_map[code] = (name, binding)
            self.controls[name] = Control(name=name, surface_id=self.id, binding=binding)

        # Parse absolute axis mappings
        self._abs_map: dict[int, tuple[str, ControlBinding, int, int]] = {}
        for code_str, axis_cfg in config.get("axes", {}).items():
            code = int(code_str)
            name = f"abs_{code}"
            binding = ControlBinding(
                action=axis_cfg.get("action", ""),
                target=axis_cfg.get("target", ""),
            )
            min_val = axis_cfg.get("min", 0)
            max_val = axis_cfg.get("max", 255)
            self._abs_map[code] = (name, binding, min_val, max_val)
            self.controls[name] = Control(name=name, surface_id=self.id, binding=binding)

        # Parse relative axis mappings (jog wheels, scroll rings)
        self._rel_map: dict[int, tuple[str, ControlBinding]] = {}
        for code_str, rel_cfg in config.get("rel_axes", {}).items():
            code = int(code_str)
            name = f"rel_{code}"
            binding = ControlBinding(
                action=rel_cfg.get("action", ""),
                target=rel_cfg.get("target", ""),
            )
            self._rel_map[code] = (name, binding)
            self.controls[name] = Control(name=name, surface_id=self.id, binding=binding)

    async def start(self) -> None:
        self._device = find_evdev_device(self._device_pattern)
        if not self._device:
            log.warning("evdev surface '%s': device matching '%s' not found",
                        self.id, self._device_pattern)
            return

        if self._grab:
            try:
                self._device.grab()
            except OSError as e:
                log.warning("Could not grab %s: %s", self._device.path, e)

        # Read actual axis info for normalization
        cap = self._device.capabilities(absinfo=True)
        if ecodes.EV_ABS in cap:
            for item in cap[ecodes.EV_ABS]:
                if isinstance(item, tuple):
                    code, absinfo = item
                    if code in self._abs_map:
                        name, binding, _, _ = self._abs_map[code]
                        self._abs_map[code] = (name, binding, absinfo.min, absinfo.max)

        self._task = asyncio.create_task(
            self._read_loop(), name=f"evdev-{self._device.path}"
        )
        log.info("evdev surface '%s' started: %s (%s) — %d buttons, %d axes, %d rel",
                 self.id, self._device.name, self._device.path,
                 len(self._button_map), len(self._abs_map), len(self._rel_map))

    async def stop(self) -> None:
        if self._task:
            self._task.cancel()
            try:
                await self._task
            except asyncio.CancelledError:
                pass

    def set_on_changed(self, callback: Any) -> None:
        self._on_changed = callback

    async def _fire(self, control_name: str, value: Any) -> None:
        if self._on_changed:
            await self._on_changed(self.id, control_name, value)

    async def _read_loop(self) -> None:
        try:
            async for event in self._device.async_read_loop():
                if event.type == ecodes.EV_KEY and event.value == 1:  # press only
                    mapping = self._button_map.get(event.code)
                    if mapping:
                        name, binding = mapping
                        val = binding.value if binding.value is not None else True
                        await self._fire(name, val)

                elif event.type == ecodes.EV_ABS:
                    mapping = self._abs_map.get(event.code)
                    if mapping:
                        name, binding, min_val, max_val = mapping
                        if max_val != min_val:
                            normalized = (event.value - min_val) / (max_val - min_val)
                        else:
                            normalized = 0.0
                        await self._fire(name, normalized)

                elif event.type == ecodes.EV_REL:
                    mapping = self._rel_map.get(event.code)
                    if mapping:
                        name, binding = mapping
                        # Relative events: positive = forward/right, negative = backward/left
                        direction = 1 if event.value > 0 else -1
                        await self._fire(name, direction)

        except asyncio.CancelledError:
            raise
        except OSError:
            log.warning("evdev device disconnected: %s", self._device.name if self._device else "?")
        except Exception as e:
            log.error("evdev surface read error: %s", e)

    def to_dict(self) -> dict:
        d = super().to_dict()
        d["device"] = self._device.name if self._device else self._device_pattern
        d["device_path"] = self._device.path if self._device else None
        return d

File: ozma-drivers/src/evdev_surface.rs
// SPDX-License-Identifier: AGPL-3.0-only
//! Config-driven evdev control surface.
//!
//! Mirrors `EvdevSurface` in `controller/evdev_surface.py`.
//!
//! Example config (JSON):
//! ```json
//! {
//!   "id": "shuttle",
//!   "device": "ShuttlePRO",
//!   "grab": true,
//!   "buttons": {
//!     "260": { "action": "scenario.next", "value": -1 }
//!   },
//!   "axes": {
//!     "7": { "action": "audio.volume", "target": "@active", "min": 0, "max": 255 }
//!   },
//!   "rel_axes": {
//!     "7": { "action": "scenario.next" }
//!   }
//! }
//! ```

use std::collections::HashMap;

use async_trait::async_trait;
use evdev::{AbsoluteAxisType, Device, InputEventKind, RelativeAxisType};
use serde_json::Value;
use tokio::sync::mpsc;
use tracing::{info, warn};

use crate::control_surface::{ControlBinding, ControlEvent, ControlInfo, ControlSurface};

// ── Device discovery ─────────────────────────────────────────────────────────

/// Find the first `/dev/input/event*` device whose name contains `pattern`
/// (case-insensitive).
///
/// Mirrors `find_evdev_device()` in `controller/evdev_surface.py`.
fn find_device(pattern: &str) -> Option<std::path::PathBuf> {
    let pattern_lower = pattern.to_lowercase();
    let Ok(entries) = std::fs::read_dir("/dev/input") else {
        return None;
    };
    for entry in entries.flatten() {
        let path = entry.path();
        if let Ok(dev) = Device::open(&path) {
            if dev
                .name()
                .unwrap_or("")
                .to_lowercase()
                .contains(&pattern_lower)
            {
                return Some(path);
            }
        }
    }
    None
}

// ── Internal mapping types ────────────────────────────────────────────────────

#[derive(Debug, Clone)]
struct ButtonMapping {
    name: String,
    binding: ControlBinding,
}

#[derive(Debug, Clone)]
struct AbsMapping {
    name: String,
    binding: ControlBinding,
    min: f64,
    max: f64,
}

#[derive(Debug, Clone)]
struct RelMapping {
    name: String,
    binding: ControlBinding,
}

// ── EvdevSurface ──────────────────────────────────────────────────────────────

/// Config-driven evdev control surface.
///
/// Mirrors `EvdevSurface` in `controller/evdev_surface.py`.
pub struct EvdevSurface {
    id: String,
    device_pattern: String,
    grab: bool,

    button_map: HashMap<u16, ButtonMapping>,
    abs_map: HashMap<u16, AbsMapping>,
    rel_map: HashMap<u16, RelMapping>,

    controls_meta: HashMap<String, ControlInfo>,

    event_tx: mpsc::Sender<ControlEvent>,
    event_rx: mpsc::Receiver<ControlEvent>,
    task: Option<tokio::task::JoinHandle<()>>,
}

impl EvdevSurface {
    /// Construct from a surface ID and JSON config object.
    pub fn new(surface_id: impl Into<String>, config: &serde_json::Value) -> Self {
        let id = surface_id.into();
        let device_pattern = config
            .get("device")
            .and_then(Value::as_str)
            .unwrap_or("")
            .to_string();
        let grab = config.get("grab").and_then(Value::as_bool).unwrap_or(false);

        let mut button_map: HashMap<u16, ButtonMapping> = HashMap::new();
        let mut abs_map: HashMap<u16, AbsMapping> = HashMap::new();
        let mut rel_map: HashMap<u16, RelMapping> = HashMap::new();
        let mut controls_meta: HashMap<String, ControlInfo> = HashMap::new();

        // Parse button mappings
        if let Some(buttons) = config.get("buttons").and_then(Value::as_object) {
            for (code_str, binding_cfg) in buttons {
                if let Ok(code) = code_str.parse::<u16>() {
                    let name = format!("btn_{code}");
                    let binding = parse_binding(binding_cfg);
                    button_map.insert(
                        code,
                        ButtonMapping { name: name.clone(), binding: binding.clone() },
                    );
                    controls_meta.insert(
                        name.clone(),
                        ControlInfo { name, surface_id: id.clone(), binding },
                    );
                }
            }
        }

        // Parse absolute axis mappings
        if let Some(axes) = config.get("axes").and_then(Value::as_object) {
            for (code_str, axis_cfg) in axes {
                if let Ok(code) = code_str.parse::<u16>() {
                    let name = format!("abs_{code}");
                    let binding = parse_binding(axis_cfg);
                    let min = axis_cfg.get("min").and_then(Value::as_f64).unwrap_or(0.0);
                    let max = axis_cfg.get("max").and_then(Value::as_f64).unwrap_or(255.0);
                    abs_map.insert(
                        code,
                        AbsMapping { name: name.clone(), binding: binding.clone(), min, max },
                    );
                    controls_meta.insert(
                        name.clone(),
                        ControlInfo { name, surface_id: id.clone(), binding },
                    );
                }
            }
        }

        // Parse relative axis mappings (jog wheels, scroll rings)
        if let Some(rel_axes) = config.get("rel_axes").and_then(Value::as_object) {
            for (code_str, rel_cfg) in rel_axes {
                if let Ok(code) = code_str.parse::<u16>() {
                    let name = format!("rel_{code}");
                    let binding = parse_binding(rel_cfg);
                    rel_map.insert(
                        code,
                        RelMapping { name: name.clone(), binding: binding.clone() },
                    );
                    controls_meta.insert(
                        name.clone(),
                        ControlInfo { name, surface_id: id.clone(), binding },
                    );
                }
            }
        }

        let (event_tx, event_rx) = mpsc::channel(256);

        Self {
            id,
            device_pattern,
            grab,
            button_map,
            abs_map,
            rel_map,
            controls_meta,
            event_tx,
            event_rx,
            task: None,
        }
    }
}

fn parse_binding(cfg: &Value) -> ControlBinding {
    ControlBinding {
        action: cfg.get("action").and_then(Value::as_str).unwrap_or("").to_string(),
        target: cfg.get("target").and_then(Value::as_str).unwrap_or("").to_string(),
        value: cfg.get("value").cloned(),
    }
}

#[async_trait]
impl ControlSurface for EvdevSurface {
    fn id(&self) -> &str {
        &self.id
    }

    fn controls(&self) -> HashMap<String, ControlInfo> {
        self.controls_meta.clone()
    }

    async fn start(&mut self) -> anyhow::Result<()> {
        let path = match find_device(&self.device_pattern) {
            Some(p) => p,
            None => {
                warn!(
                    "evdev surface '{}': device matching '{}' not found",
                    self.id, self.device_pattern
                );
                return Ok(());
            }
        };

        let mut dev = Device::open(&path)?;

        if self.grab {
            if let Err(e) = dev.grab() {
                warn!("Could not grab {:?}: {}", path, e);
            }
        }

        // Refine ABS axis ranges from actual device absinfo
        if let Ok(abs_state) = dev.get_absinfo() {
            for (code, mapping) in self.abs_map.iter_mut() {
                if let Some(info) = abs_state.get(*code as usize) {
                    mapping.min = info.minimum() as f64;
                    mapping.max = info.maximum() as f64;
                }
            }
        }

        let surface_id = self.id.clone();
        let button_map = self.button_map.clone();
        let abs_map = self.abs_map.clone();
        let rel_map = self.rel_map.clone();
        let tx = self.event_tx.clone();

        info!(
            "evdev surface '{}' started: {} ({:?}) — {} buttons, {} axes, {} rel",
            self.id,
            dev.name().unwrap_or("?"),
            path,
            button_map.len(),
            abs_map.len(),
            rel_map.len(),
        );

        let task = tokio::spawn(async move {
            let mut stream = match dev.into_event_stream() {
                Ok(s) => s,
                Err(e) => {
                    warn!("evdev surface '{}': failed to open event stream: {}", surface_id, e);
                    return;
                }
            };

            loop {
                let raw = match stream.next_event().await {
                    Ok(e) => e,
                    Err(e) => {
                        warn!("evdev surface '{}' read error: {}", surface_id, e);
                        break;
                    }
                };

                let evt: Option<ControlEvent> = match raw.kind() {
                    // Button press only (value == 1); ignore release and repeat
                    InputEventKind::Key(key) if raw.value() == 1 => {
                        let code = key.code();
                        button_map.get(&code).map(|m| {
                            let val = m.binding.value.clone().unwrap_or(Value::Bool(true));
                            ControlEvent {
                                surface_id: surface_id.clone(),
                                control_name: m.name.clone(),
                                value: val,
                                binding: m.binding.clone(),
                            }
                        })
                    }
                    InputEventKind::AbsAxis(axis) => {
                        let code = axis.0;
                        abs_map.get(&code).map(|m| {
                            let normalized = if (m.max - m.min).abs() > f64::EPSILON {
                                (raw.value() as f64 - m.min) / (m.max - m.min)
                            } else {
                                0.0
                            };
                            ControlEvent {
                                surface_id: surface_id.clone(),
                                control_name: m.name.clone(),
                                value: Value::from(normalized),
                                binding: m.binding.clone(),
                            }
                        })
                    }
                    InputEventKind::RelAxis(axis) => {
                        let code = axis.0;
                        rel_map.get(&code).map(|m| {
                            // Positive = forward/right, negative = backward/left
                            let direction: i64 = if raw.value() > 0 { 1 } else { -1 };
                            ControlEvent {
                                surface_id: surface_id.clone(),
                                control_name: m.name.clone(),
                                value: Value::from(direction),
                                binding: m.binding.clone(),
                            }
                        })
                    }
                    _ => None,
                };

                if let Some(event) = evt {
                    if tx.send(event).await.is_err() {
                        break; // receiver dropped
                    }
                }
            }
        });

        self.task = Some(task);
        Ok(())
    }

    async fn stop(&mut self) -> anyhow::Result<()> {
        if let Some(task) = self.task.take() {
            task.abort();
            let _ = task.await;
        }
        Ok(())
    }

    async fn next_event(&mut self) -> Option<ControlEvent> {
        self.event_rx.recv().await
    }

    fn to_dict(&self) -> serde_json::Value {
        serde_json::json!({
            "id": self.id,
            "device_pattern": self.device_pattern,
            "controls": self.controls_meta.keys().collect::<Vec<_>>(),
        })
    }
}
